File: samplesJS/GamePads/src/GamePadsSample.cpp
// Required for WinMain
#include <Windows.h>

// Oak Nut include
#include <onut/GamePad.h>
#include <onut/onut.h>
#include <onut/Renderer.h>
#include <onut/Settings.h>
#include <onut/SpriteBatch.h>
#include <onut/Texture.h>

void init();
void update();
void render();

// Main
int CALLBACK WinMain(HINSTANCE appInstance, HINSTANCE prevInstance, LPSTR cmdLine, int cmdCount)
{
    oSettings->setGameName("Sprites Sample");
    ORun(init, update, render);
}

void init()
{
}

void update()
{
}

void drawController(int index, const Vector2& position)
{
    if (!OGetGamePad(index)->isConnected())
    {
        oSpriteBatch->drawSprite(OGetTexture("gamepad.png"), position, Color(.5f));
    }
    else
    {
        oSpriteBatch->drawSprite(OGetTexture(("gamepad" + std::to_string(index) + ".png").c_str()), position);
        oSpriteBatch->drawSprite(OGetTexture("lthumb.png"), position + OGetGamePadLeftThumb(index) * 12.f);
        oSpriteBatch->drawSprite(OGetTexture("rthumb.png"), position + OGetGamePadRightThumb(index) * 12.f);
        if (OGamePadPressed(OGamePadLeftThumb, index)) oSpriteBatch->drawSprite(OGetTexture("lthumbPress.png"), position + OGetGamePadLeftThumb(index) * 12.f);
        if (OGamePadPressed(OGamePadRightThumb, index)) oSpriteBatch->drawSprite(OGetTexture("rthumbPress.png"), position + OGetGamePadRightThumb(index) * 12.f);
        if (OGamePadPressed(OGamePadRightBumper, index)) oSpriteBatch->drawSprite(OGetTexture("rb.png"), position);
        if (OGamePadPressed(OGamePadLeftBumper, index)) oSpriteBatch->drawSprite(OGetTexture("lb.png"), position);
        if (OGamePadPressed(OGamePadBack, index)) oSpriteBatch->drawSprite(OGetTexture("back.png"), position);
        if (OGamePadPressed(OGamePadStart, index)) oSpriteBatch->drawSprite(OGetTexture("start.png"), position);
        if (OGamePadPressed(OGamePadA, index)) oSpriteBatch->drawSprite(OGetTexture("a.png"), position);
        if (OGamePadPressed(OGamePadB, index)) oSpriteBatch->drawSprite(OGetTexture("b.png"), position);
        if (OGamePadPressed(OGamePadX, index)) oSpriteBatch->drawSprite(OGetTexture("x.png"), position);
        if (OGamePadPressed(OGamePadY, index)) oSpriteBatch->drawSprite(OGetTexture("y.png"), position);
        if (OGamePadPressed(OGamePadDPadLeft, index)) oSpriteBatch->drawSprite(OGetTexture("dleft.png"), position);
        if (OGamePadPressed(OGamePadDPadRight, index)) oSpriteBatch->drawSprite(OGetTexture("dright.png"), position);
        if (OGamePadPressed(OGamePadDPadUp, index)) oSpriteBatch->drawSprite(OGetTexture("dup.png"), position);
        if (OGamePadPressed(OGamePadDPadDown, index)) oSpriteBatch->drawSprite(OGetTexture("ddown.png"), position);
        if (OGamePadPressed(OGamePadLeftTrigger, index)) oSpriteBatch->drawSprite(OGetTexture("lt.png"), position);
        if (OGamePadPressed(OGamePadRightTrigger, index)) oSpriteBatch->drawSprite(OGetTexture("rt.png"), position);
    }
}

void render()
{
    // Clear
    oRenderer->clear(Color::Black);

    // Begin a batch
    oSpriteBatch->begin();

    drawController(0, Vector2(OScreenWf * .3f, OScreenHf * .3f));
    drawController(1, Vector2(OScreenWf * .7f, OScreenHf * .3f));
    drawController(2, Vector2(OScreenWf * .3f, OScreenHf * .7f));
    drawController(3, Vector2(OScreenWf * .7f, OScreenHf * .7f));

    // End and flush the batch
    oSpriteBatch->end();
}

File: samplesJS/GamePads/assets/scripts/GamePadsSample.js
var gamepads = [
    getTexture("gamepad0.png"),
    getTexture("gamepad1.png"),
    getTexture("gamepad2.png"),
    getTexture("gamepad3.png")
];
var gamepad = getTexture("gamepad.png");
var lthumb = getTexture("lthumb.png");
var rthumb = getTexture("rthumb.png");
var lthumbPress = getTexture("lthumbPress.png");
var rthumbPress = getTexture("rthumbPress.png");
var rb = getTexture("rb.png");
var lb = getTexture("lb.png");
var back = getTexture("back.png");
var start = getTexture("start.png");
var a = getTexture("a.png");
var b = getTexture("b.png");
var x = getTexture("x.png");
var y = getTexture("y.png");
var dleft = getTexture("dleft.png");
var dright = getTexture("dright.png");
var dup = getTexture("dup.png");
var ddown = getTexture("ddown.png");
var lt = getTexture("lt.png");
var rt = getTexture("rt.png");

function drawController(index, position) {
    if (!GamePad.isConnected(index)) {
        SpriteBatch.drawSprite(gamepad, position, new Color(.5));
    }
    else {
        SpriteBatch.drawSprite(gamepads[index], position);
        SpriteBatch.drawSprite(lthumb, GamePad.getLeftThumb(index).mul(12).add(position));
        SpriteBatch.drawSprite(rthumb, GamePad.getRightThumb(index).mul(12).add(position));
        if (GamePad.isDown(index, Button.LEFT_THUMBSTICK)) SpriteBatch.drawSprite(lthumbPress, GamePad.getLeftThumb(index).mul(12).add(position));
        if (GamePad.isDown(index, Button.RIGHT_THUMBSTICK)) SpriteBatch.drawSprite(rthumbPress, GamePad.getRightThumb(index).mul(12).add(position));
        if (GamePad.isDown(index, Button.RIGHT_BUMPER)) SpriteBatch.drawSprite(rb, position);
        if (GamePad.isDown(index, Button.LEFT_BUMPER)) SpriteBatch.drawSprite(lb, position);
        if (GamePad.isDown(index, Button.BACK)) SpriteBatch.drawSprite(back, position);
        if (GamePad.isDown(index, Button.START)) SpriteBatch.drawSprite(start, position);
        if (GamePad.isDown(index, Button.A)) SpriteBatch.drawSprite(a, position);
        if (GamePad.isDown(index, Button.B)) SpriteBatch.drawSprite(b, position);
        if (GamePad.isDown(index, Button.X)) SpriteBatch.drawSprite(x, position);
        if (GamePad.isDown(index, Button.Y)) SpriteBatch.drawSprite(y, position);
        if (GamePad.isDown(index, Button.DPAD_LEFT)) SpriteBatch.drawSprite(dleft, position);
        if (GamePad.isDown(index, Button.DPAD_RIGHT)) SpriteBatch.drawSprite(dright, position);
        if (GamePad.isDown(index, Button.DPAD_UP)) SpriteBatch.drawSprite(dup, position);
        if (GamePad.isDown(index, Button.DPAD_DOWN)) SpriteBatch.drawSprite(ddown, position);
        if (GamePad.isDown(index, Button.LEFT_TRIGGER)) SpriteBatch.drawSprite(lt, position);
        if (GamePad.isDown(index, Button.RIGHT_TRIGGER)) SpriteBatch.drawSprite(rt, position);
    }
}

function render() {
    // Clear
    Renderer.clear(Color.BLACK);

    // Begin a batch
    SpriteBatch.begin();

    var screenResolution = Renderer.getResolution();
    drawController(0, new Vector2(screenResolution.x * .3, screenResolution.y * .3));
    drawController(1, new Vector2(screenResolution.x * .7, screenResolution.y * .3));
    drawController(2, new Vector2(screenResolution.x * .3, screenResolution.y * .7));
    drawController(3, new Vector2(screenResolution.x * .7, screenResolution.y * .7));

    // End and flush the batch
    SpriteBatch.end();
}
